src/mozanalysis/experiment.py

```diff
@@ -607,6 +609,60 @@ def create(
         )
         return tl
 
+    @classmethod
+    def for_ts(
```

Comment:
Let's explore "everything is a timeseries" a little:
1) `'single'` does not adequately specify a non-time-series analysis window: `analysis_start_days` and `analysis_length_dates` must also be supplied iff `time_series_period='single'`.
2) At present, `get_per_client_data()` returns a pyspark DataFrame in "the standard format", and `get_time_series_data()` returns a dict of pandas DataFrames, each of which are in "the standard format". If we switch to one method only, then would it return a one-element dict? That seems fairly unergonomic? Or would we make it "unpredictable" instead (which is unlikely to annoy any user for more than 20 sec, once in their life - though it will make the docs more complicated?)
3) I had been hanging onto the untested theory that the spark query would be faster if the `days_since_enrollment.between()` condition was defined w.r.t. int literals rather than columns, but I think I should let this go as at best premature optimisation and at worst superstition.

I think (1) is the only one of these to take seriously?

Reply:
1) The concern is that we'd need the constructor to accept additional arguments in that case, right? If we wanted to to allow for analysis windows that were not "the entire experiment"? In that case I agree my suggestion is bad, but this could still produce a timeseries-style limits object with a second constructor, right?
2) One-element dict is my favorite of these two (consistency!) though I suggested elsewhere that maybe the return value of get_time_series_data() should be a single spark dataframe in tidy format with (client_id, analysis_window) ID columns, which I also like.

Reply:
Re 1.: Given that the historical/current definition of retention is something like "a ping received in the third week after enrolling" - yes we do need arbitrary analysis windows.

> this could still produce a timeseries-style limits object with a second constructor, right?

Yes, we _could_ add an `analysis_window_start` column to the `enrollments` dataframe even when not querying a time series of data and thus avoid the `if time_limits.time_series_period is None` branch in `_get_join_conditions()`. Though `TimeLimits` would still need to know about `analysis_window_start` or we'd need to find another way to pass that info into `_process_enrollments()`, which would be messy. So the gross improvement would be the removal of a branch condition and the potential removal of the attribute `TimeLimits.analysis_window_end` (though I'd like to keep it, for validation)

> One-element dict is my favorite of these two (consistency!)

If "the standard format" accepted by the stats functions remains "a DataFrame with one row per client and one column with branch name and one column with metric data", then we should return data in the standard format wherever possible, so that it as easy as possible to use the stats functions (no searching around for your keys in the single metric case!). If we change "the standard format" to be a _dictionary_ of such dataframes and update the stats functions accordingly then it would make sense to return a one-element dict - though this of course has the downside of making it inconvenient to run computationally expensive stats for just a single point of the time series (I elaborate more on this argument below)

> I suggested elsewhere that maybe the return value of get_time_series_data() should be a single spark dataframe in tidy format with (client_id, analysis_window) ID columns, which I also like.

I think I said verbally that I consider this a footgun, because the stats functions assume that there is exactly one row per client, and have no way of checking whether this is true; it would be very easy for a DS to forget to filter the DataFrame to a single analysis window before feeding it to the stats functions. Alternatively we could  modify "the standard format" to include an analysis window column, and modify the stats functions to iterate over the sets of rows with shared analysis windows? But I think this is a bad direction: when possible we should prefer to define a function `X -> Y` then map a list over it to obtain `[X] -> [Y]` when needed, rather than commit to a direction where the simplest function we can define is `[X] -> [Y]`? Is this convincing to you?

Reply:
> the removal of a branch condition

I count two -- maybe "a lot" was overselling it :)

> Alternatively we could modify "the standard format" to include an analysis window column, and modify the stats functions to iterate over the sets of rows with shared analysis windows?

Yeah, I was enthusiastic about this when you suggested it.

> But I think this is a bad direction: when possible we should prefer to define a function X -> Y then map a list over it to obtain [X] -> [Y] when needed, rather than commit to a direction where the simplest function we can define is [X] -> [Y]? Is this convincing to you?

Not on its face; I have different aesthetic preferences. With respect to the user interface, I think options and special cases are bad. They increase complexity, require users and developers to make more decisions and incur the related cognitive load, and cause type confusions. I think it's okay to be a little verbose to be consistent. I'd favor expressing the API in terms of the general case, especially since the general case is also pretty common.

> though [a dictionary of such dataframes] of course has the downside of making it inconvenient to run computationally expensive stats for just a single point of the time series

Bleh, right, subsets of dictionaries are not dictionaries. That's annoying.

I think we're not likely to convince each other; your approach here seems reasonable even if I don't prefer it so I don't think you should treat my comments here as blocking.

Reply:
OK; thanks for the "disagree and commit" re: whether the time series vs one analysis window usecases are separate enough to warrant specialised function signatures. In parting I'll note that if mozreport wants a consistent API it should be simple to wrap `get_per_client_data()` to make it consistent within the mozreport context.

Reply:
I think I've arrived at a solution that will make us both happy. We should keep my proposed interface, but make lots of internal changes in the spirit of "everything is a time series" (actually it will be "everything has an arbitrary list of analysis windows"). This gets hidden from people calling the proposed methods but you could call `_get_per_client_data()` directly with a `time_limits` object you created.

So humans can hack with the convenience methods specific to what they're looking at, while you get direct access to something consistent, powerful, and able to do lots of things that no-one is interested in.

Reply:
N.B. by "able to do lots of things that no-one is interested in" I didn't mean "this is pointless"; I meant "able to do lots of things that no-one is interested in as well as all the things people are interested in", a.k.a. it's a lower level interface